### ISSUE TITLE
bcrypt: Problem with dupe hashes

### ISSUE THREAD

**Issue by magnumripper** [openwall/john#4390]
See #4388 

This issue is about vec3/vec11 (identical entries) and not necessarily just with bcrypt.

John Proper: (**EDIT: this was a non-OMP build**)
```
$ ln -fs ../test/tst-.pot ../run/john.pot 
$ echo -n > tst-.pot && ../run/john -ses=tst-  tst-.in --wordlist=tst-.dic -form=bcrypt
Loaded 13 password hashes with 6 different salts (bcrypt [Blowfish 32/64 X3])
Press 'q' or Ctrl-C to abort, almost any other key for status
U*U              (vec_0)
U*U*             (vec_1)
U*U*U            (vec_13)
U*U*U            (vec_2)
                 (vec_3)
?                (vec_6)
?                (vec_5)
0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 (vec_4)
???????????????????????????????????????????????????????????????????????? (vec_9)
??????           (vec_8)
ё               (vec_7)
?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U (vec_10)
U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U?? (vec_12)
13g 0:00:00:00 100% 325.0g/s 450.0p/s 1950c/s 3975C/s ?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U..U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??
Warning: passwords printed above might not be all those cracked
Use the "--show" option to display all of the cracked passwords reliably
Session completed
```
Loaded 13, cracked 13. Only 12 of them ended up in the pot file but that is #4388 and irrelevant for this issue.

Jumbo:
```
$ rm -f tst-.pot && ../run/john -ses=tst-  -skip -pot=tst-.pot tst-.in --wordlist=tst-.dic -form=bcrypt 
Using default input encoding: UTF-8
Loaded 13 password hashes with 6 different salts (2.2x same-salt boost) (bcrypt [Blowfish 32/64 X3])
Cost 1 (iteration count) is 32 for all loaded hashes
Will run 16 OpenMP threads
Press 'q' or Ctrl-C to abort, almost any other key for status
Warning: Only 19 candidates left, minimum 48 needed for performance.
U*U              (vec_0)
U*U*             (vec_1)
                 (vec_3)
                 (vec_3)
U*U*U            (vec_13)
U*U*U            (vec_2)
0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 (vec_4)
?                (vec_5)
?                (vec_6)
???????????????????????????????????????????????????????????????????????? (vec_9)
?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U?U (vec_10)
U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??U??               (vec_12)
??????              (vec_8)
ё                (vec_7)
14g 0:00:00:00 DONE (2020-09-30 16:17) 700.0g/s 950.0p/s 5700c/s 12350C/s 123456789012345678901234567890123456789012345678901234567890123456789012..U*U*U
Warning: passwords printed above might not be all those cracked
Use the "--show" option to display all of the cracked passwords reliably
Session completed. 
```
Loaded 13, cracked 14(!) and the vec_3/vec_11 hash ended up twice in the pot file.

**Comment by magnumripper** [comment]
Doesn't happen with NT format. Perhaps I should try a salted one.

**Comment by magnumripper** [comment]
I can't even replicate the bug with bcrypt, using a few non-test vector words. Need to accomplish that first. I wonder what exactly is triggering this?!

**Comment by magnumripper** [comment]
This was present at least as long ago as Aug 2015, #1607! I was mislead by the vec2 issue and never noticed we differ from John Proper handling vec3/vec11.

**Comment by magnumripper** [comment]
Off-topic notes to self. Trying to build an old version of our code on macOS (during bisecting), for some reason OpenSSL isn't found with the usual
```
./configure LDFLAGS=-L/usr/local/opt/openssl/lib CPPFLAGS=-I/usr/local/opt/openssl/include
```
stuff, but
```
./configure PKG_CONFIG_PATH=/usr/local/opt/openssl/lib/pkgconfig
```
worked fine. But then `make` fail on everything because my OpenSSL is too new (they really went through hoops ensuring not to be backward compatible). Perhaps the configure problem is also a version thing, I can't remember.

**Comment by claudioandre-br** [comment]
I have to use in some circumstances:
```
"OPENSSL_CFLAGS=-I/usr/local/opt/openssl@1.1/include OPENSSL_LIBS='-L/usr/local/Cellar/openssl@1.1/1.1.1d/lib -lssl -lcrypto'"

**Comment by magnumripper** [comment]
I think what I need is Homebrew openssl@1.0 but it doesn't seem to be available.

**Comment by magnumripper** [comment]
Here's an interesting data point: Using `OMP_NUM_THREADS=1` does not make the bug go away - but I have yet to see it happen with a non-OMP build!

On a side note, #4388 is not affected the same way, it still shows.

**Comment by magnumripper** [comment]
I confirmed that Jumbo's code, while moved around due to having `BF_common*` shared with OpenCL, is essentially pristine. So it's not a Jumbo-only bcrypt-only bug. Should be something with `cracker.c`.

But `cracker.c` has *no* OpenMP code, so the bug only showing with OpenMP is a strange fact.

**Comment by magnumripper** [comment]
Oh, and I realized I should re-test John proper, using OpenMP. This too **is** a bug in it!

**Comment by magnumripper** [comment]
I have yet to trigger the bug using anything else than the wordlist TS created (from test vectors + random garbage). Here it is:

[BF_hashes.txt](https://github.com/openwall/john/files/5315154/BF_hashes.txt)
[BF_TS_wordlist.txt](https://github.com/openwall/john/files/5315155/BF_TS_wordlist.txt)

**Comment by magnumripper** [comment]
The segfault in #4389 is indeed just an effect of this bug: After the "14 guesses" out of 13 possible, quite consequently `db->password_count` is `-1` in the end of the run so using `ShowRemainOnStatus = Y` we enter code that assumes `db->salt_count` must be non-zero, which it's not. So division by zero.

**Comment by solardiz** [comment]
Reduced test case that still triggers duplicate logging of the `vec_3` guess:

```
vec_0:$2a$05$CCCCCCCCCCCCCCCCCCCCC.E5YPO9kmyuRGyh0XouQYb4YMJKvyOeW
vec_1:$2a$05$CCCCCCCCCCCCCCCCCCCCC.VGOzA784oUp/Z0DY336zx7pLYAy0lwK
vec_2:$2a$05$XXXXXXXXXXXXXXXXXXXXXOAcXxm9kjPGEMsLznoKqmqw7tc8WCx4a
vec_3:$2a$05$CCCCCCCCCCCCCCCCCCCCC.7uG0VCzI2bS7j6ymqJi9CdcdxiRTWNy
```

```
U*U
U*U*

x

```

**Comment by solardiz** [comment]
Here's one way to fix this:

```diff
+++ john-1.9.0/src/cracker.c	2020-10-29 21:29:53 +0100
@@ -49,6 +49,7 @@ static void *crk_last_salt;
 static void (*crk_fix_state)(void);
 static struct db_keys *crk_guesses;
 static uint64_t *crk_timestamps;
+static void *crk_removed_hash;
 static char crk_stdout_key[PLAINTEXT_BUFFER_SIZE];
 
 static void crk_dummy_set_salt(void *salt)
@@ -132,6 +133,7 @@ void crk_init(struct db_main *db, void (
 	if (db->loaded) {
 		size = crk_params.max_keys_per_crypt * sizeof(uint64_t);
 		memset(crk_timestamps = mem_alloc(size), -1, size);
+		crk_removed_hash = NULL;
 	} else
 		crk_stdout_key[0] = 0;
 
@@ -159,6 +161,21 @@ static void crk_remove_salt(struct db_sa
 }
 
 /*
+ * Marks the hash as removed by setting its binary pointer to a magic value.
+ * We don't simply set it to NULL to avoid having to pre-check for NULL in
+ * performance-critical places.  We prefer to post-check for the magic value
+ * on the occasional duplicate or even completely spurious guesses.
+ */
+static void crk_mark_hash_as_removed(struct db_password *pw)
+{
+	if (!crk_removed_hash) {
+		crk_removed_hash = mem_alloc(crk_params.binary_size);
+		memcpy(crk_removed_hash, pw->binary, crk_params.binary_size);
+	}
+	pw->binary = crk_removed_hash;
+}
+
+/*
  * Updates the database after a password has been cracked.
  */
 static void crk_remove_hash(struct db_salt *salt, struct db_password *pw)
@@ -171,6 +188,7 @@ static void crk_remove_hash(struct db_sa
 	if (!--salt->count) {
 		salt->list = NULL; /* "single crack" mode might care */
 		crk_remove_salt(salt);
+		crk_mark_hash_as_removed(pw);
 		return;
 	}
 
@@ -234,6 +252,9 @@ static int crk_process_guess(struct db_s
 	int dupe;
 	char *key;
 
+	if (!pw->binary || pw->binary == crk_removed_hash) /* Already logged */
+		return 0;
+
 	dupe = crk_timestamps[index] == status.crypts;
 	crk_timestamps[index] = status.crypts;
 
@@ -589,5 +610,6 @@ void crk_done(void)
 			crk_salt_loop();
 
 		MEM_FREE(crk_timestamps);
+		MEM_FREE(crk_removed_hash);
 	}
 }
```

I think I can make it simpler - still trying.

**Comment by solardiz** [comment]
Looks like it's as simple as this:

```diff
+++ john-1.9.0/src/cracker.c	2020-10-29 21:49:09 +0100
@@ -171,7 +171,8 @@ static void crk_remove_hash(struct db_sa
 	if (!--salt->count) {
 		salt->list = NULL; /* "single crack" mode might care */
 		crk_remove_salt(salt);
-		return;
+		if (!salt->bitmap)
+			return;
 	}
 
 /*
```

In plain English, when removing the last hash for a salt, we didn't bother updating the bitmap and hash table for the salt. Now we will (unfortunately, at some increase of memory consumption with "--fork" because of copy-on-write). I was worried this might not fully fix the problem when we're prefetching, but apparently I had already implemented logic to deal with that:

```c
/* After we've successfully cracked and removed a hash, our prefetched bitmap
 * and hash table entries might be stale: some might correspond to the same
 * hash bucket, yet with this removed hash still in there if it was the first
 * one in the bucket.  If so, re-prefetch from the next lucky index if any,
 * yet complete handling of this index first. */
```

The "complete handling of this index first" part should be safe as it's going through `hash_next` pointers, which cannot hit the same hash again.

**Comment by solardiz** [comment]
The main issue reported here, and the one I deal with so far, has nothing to do with bcrypt and its peculiarities, nor even with the general possibility of having differing `source` along with matching `binary` and `salt` - indeed, my reduced test case doesn't have any of that (it only has very explicitly matching salts, but no matching hashes).

However, while playing with this I see that we do indeed also have other issues that are exposed when binaries match yet sources don't, or/and when binary salts match yet sources don't. This is again not specific to just bcrypt.

**Comment by solardiz** [comment]
@magnumripper I notice that in jumbo that short-circuit `return;` also skipped `BLOB_FREE`. With my patch above, it will now be reached when there's a bitmap, but will still be skipped without a bitmap when removing the last hash for a salt. Please comment on this.

Also, should the `BLOB_FREE` be conditional on `FMT_REMOVE` or not? Right now, the two places in `crk_remove_hash` are inconsistent in this respect.

**Comment by solardiz** [comment]
> Also, should the `BLOB_FREE` be conditional on `FMT_REMOVE` or not?

I don't see a correlation between uses of `FMT_BLOB` and `FMT_REMOVE`. I understand it feels wrong to be freeing the blob without setting the `binary` pointer to `NULL`.

**Comment by magnumripper** [comment]
> Also, should the BLOB_FREE be conditional on FMT_REMOVE or not? Right now, the two places in crk_remove_hash are inconsistent in this respect.

I'm nearly certain we should always `BLOB_FREE(pw->binary)` before setting `pw->binary` to NULL, so `FMT_REMOVE` should affect both or none. That's consistent.

**Comment by magnumripper** [comment]
> @magnumripper I notice that in jumbo that short-circuit `return;` also skipped `BLOB_FREE`. With my patch above, it will now be reached when there's a bitmap, but will still be skipped without a bitmap when removing the last hash for a salt. Please comment on this.

I need to stare at that for a while but that may be a minor bug (the only result being we don't free up memory we could free up).

**Comment by solardiz** [comment]
> I'm nearly certain we should always `BLOB_FREE(pw->binary)` before setting `pw->binary` to NULL, so `FMT_REMOVE` should affect both or none. That's consistent.

We don't always set `pw->binary` to NULL when we're removing that hash as cracked. We sometimes skip that to avoid NULL pointer dereference or having to pre-check for NULL in the cracking loops. We also sometimes skip that to minimize memory waste from copy-on-write with "--fork". Yet in those cases we might nevertheless want to free the blob, if any. I don't see how `FMT_REMOVE` is relevant.

Also, when using blobs we probably have few enough loaded that we don't care about the slight waste from copy-on-write of individual pages on removal.

> a minor bug (the only result being we don't free up memory we could free up).

Sure.

**Comment by magnumripper** [comment]
> We don't always set `pw->binary` to NULL when we're removing that hash as cracked. We sometimes skip that to avoid NULL pointer dereference or having to pre-check for NULL in the cracking loops. We also sometimes skip that to minimize memory waste from copy-on-write with "--fork". Yet in those cases we might nevertheless want to free the blob, if any.

Definitely. I guess I simply couldn't identify those cases when I implemented it. I'll create a PR. **edit:** #4432
